components/server/src/auth/auth-provider-service.ts

```diff
@@ -56,77 +63,279 @@ export class AuthProviderService {
+    private isVerified = (info: AuthProviderInfo | AuthProviderParams) => info.verified;
+    private isNotOrgProvider = (info: AuthProviderInfo | AuthProviderParams) => !info.organizationId;
+
+    async getAuthProviderDescriptionsUnauthenticated(): Promise<AuthProviderInfo[]> {
+        const { builtinAuthProvidersConfigured } = this.config;
+
+        const authProviders = [...(await this.getAllAuthProviderParams()), ...this.config.authProviderConfigs];
+
+        const toPublic = (ap: AuthProviderParams) =>
+            <AuthProviderInfo>{
+                authProviderId: ap.id,
+                authProviderType: ap.type,
+                disallowLogin: ap.disallowLogin,
+                host: ap.host,
+                icon: ap.icon,
+                description: ap.description,
+            };
+        let result = authProviders.filter(this.isNotHidden).filter(this.isVerified).filter(this.isNotOrgProvider);
+        if (builtinAuthProvidersConfigured) {
+            result = result.filter(this.isBuiltIn);
+        }
+        return result.map(toPublic);
+    }
+
+    async getAuthProviderDescriptions(user: User): Promise<AuthProviderInfo[]> {
+        const { builtinAuthProvidersConfigured } = this.config;
+
+        const authProviders = [...(await this.getAllAuthProviderParams()), ...this.config.authProviderConfigs];
+
+        // explicitly copy to avoid bleeding sensitive details
+        const toInfo = (ap: AuthProviderParams) =>
+            <AuthProviderInfo>{
+                authProviderId: ap.id,
+                authProviderType: ap.type,
+                ownerId: ap.ownerId,
+                organizationId: ap.organizationId,
+                verified: ap.verified,
+                host: ap.host,
+                icon: ap.icon,
+                hiddenOnDashboard: ap.hiddenOnDashboard,
+                disallowLogin: ap.disallowLogin,
+                description: ap.description,
+                scopes: this.toScopes(ap),
+                requirements: this.toScopeRequirements(ap),
+            };
+
+        const result: AuthProviderInfo[] = [];
+        for (const p of authProviders) {
+            const identity = user.identities.find((i) => i.authProviderId === p.id);
+            if (identity) {
+                result.push(toInfo(p));
+                continue;
+            }
+            if (p.ownerId === user.id) {
+                result.push(toInfo(p));
+                continue;
+            }
+            if (builtinAuthProvidersConfigured && !this.isBuiltIn(p)) {
+                continue;
+            }
+            if (this.isNotHidden(p) && this.isVerified(p)) {
+                result.push(toInfo(p));
+            }
+        }
         return result;
     }
+    /**
+     * This is extracted from auth provider handlers in order to avoid a
+     * dependency to host contexts.
+     *
+     * TODO(at) these defaults of provider types needs to be centralized.
```

Comment:
Does it make sense to introduce a new class used by both and move it there? Can break host contexts?

Reply:
> Can break host contexts?

no.

> Does it make sense to introduce a new class used by both and move it there? 

Currently this is done in runtime handlers of the providers (e.g. GitHubAuthProvider) and thus we were accessing this information via host context as the runtime handler is attached to it. 

The problem is, currently the runtime handlers are created per host, that comes with a lot of problems. The code dependency is just one of them. Most problems would be solved by creating those handlers per _type_ (e.g. `GitHub`).

That said, the suggestion to extract to a new class to make it reusable sounds good. I'm not sure though if need to make it in scope of this PR, or when we're to refactor auth provider handlers.

Reply:
done in 9be91ec75c375ced95c446a21a4d51dbdbcd4427